In [1]:
m = tf.constant(42)
m

NameError: name 'tf' is not defined

In [ ]:
ineuron.numpy()

42

In [2]:
pip install tensorflow

In [ ]:
ineuron1 = tf.constant(1, dtype = tf.int64)
ineuron1

<tf.Tensor: shape=(), dtype=int64, numpy=1>

In [ ]:
ineuron_x = tf.constant([[4,2],[9,5]])
print(ineuron_x)

tf.Tensor(
[[4 2]
 [9 5]], shape=(2, 2), dtype=int32)


In [ ]:
ineuron_x.numpy()

array([[4, 2],
       [9, 5]], dtype=int32)

In [ ]:
print('shape:',ineuron_x.shape)
print(ineuron_x.dtype)

shape: (2, 2)
<dtype: 'int32'>


#### Commonly used method is to generate constant tf.ones and the tf.zeros like of numpy np.ones & np.zeros

In [ ]:
print(tf.ones(shape=(2,3)))

tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]], shape=(2, 3), dtype=float32)


In [ ]:
print(tf.zeros(shape=(3,2)))

tf.Tensor(
[[0. 0.]
 [0. 0.]
 [0. 0.]], shape=(3, 2), dtype=float32)


In [ ]:
import tensorflow as tf

const2 = tf.constant([[3,4,5], [3,4,5]])
const1 = tf.constant([[1,2,3], [1,2,3]])
result = tf.add(const1, const2)

print(result)

tf.Tensor(
[[4 6 8]
 [4 6 8]], shape=(2, 3), dtype=int32)


>We have defined two constants and we add one value to the other.
>As a result, we got a Tensor object with the result of the adding.

#### Random constant

In [ ]:
tf.random.normal(shape=(2,2),mean=0,stddev=1.0)

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[-0.37854993,  0.7752413 ],
       [ 1.4218645 ,  1.1185031 ]], dtype=float32)>

In [ ]:
tf.random.uniform(shape=(2,2),minval=0,maxval=10,dtype=tf.int32)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[7, 6],
       [7, 3]], dtype=int32)>

In [ ]:
var0 = 24 # python variable
var1 = tf.Variable(42) # rank 0 tensor
var2 = tf.Variable([ [ [0., 1., 2.], [3., 4., 5.] ], [ [6., 7., 8.], [9., 10., 11.] ] ]) #rank 3 tensor
var0, var1, var2

(24,
 <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=42>,
 <tf.Variable 'Variable:0' shape=(2, 2, 3) dtype=float32, numpy=
 array([[[ 0.,  1.,  2.],
         [ 3.,  4.,  5.]],
 
        [[ 6.,  7.,  8.],
         [ 9., 10., 11.]]], dtype=float32)>)

In [ ]:
float_var64 = tf.Variable(89, dtype = tf.float64)
float_var64.dtype

tf.float64

`TensorFlow has a large number of built-in datatypes.`

datatype | description
- | -
 tf.float16 | 16-bit half-precision floating-point.
 tf.float32 | 32-bit single-precision floating-point.
 tf.float64 | 64-bit double-precision floating-point.
 tf.bfloat16 | 16-bit truncated floating-point.
 tf.complex64 | 64-bit single-precision complex.
 tf.complex128 | 128-bit double-precision complex.
 tf.int8 | 8-bit signed integer.
 tf.uint8 | 8-bit unsigned integer.
 tf.uint16 | 16-bit unsigned integer.
 tf.uint32 | 32-bit unsigned integer.
 tf.uint64 | 64-bit unsigned integer.
 tf.int16 | 16-bit signed integer.
 tf.int32 | 32-bit signed integer.
 tf.int64 | 64-bit signed integer.
 tf.bool | Boolean.
 tf.string | String.
 tf.qint8 | Quantized 8-bit signed integer.
 tf.quint8 | Quantized 8-bit unsigned integer.
 tf.qint16 | Quantized 16-bit signed integer.
 tf.quint16 | Quantized 16-bit unsigned integer.
 tf.qint32 | Quantized 32-bit signed integer.
 tf.resource | Handle to a mutable resource.
 tf.variant | Values of arbitrary types.


#### To reassign a variable, use var.assign()

In [ ]:
var_reassign = tf.Variable(89.)
var_reassign

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=89.0>

In [ ]:
var_reassign.assign(98.)
var_reassign

<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=98.0>

In [ ]:
initial_value = tf.random.normal(shape=(2,2))
a = tf.Variable(initial_value)
print(a)

<tf.Variable 'Variable:0' shape=(2, 2) dtype=float32, numpy=
array([[ 1.007538  ,  0.90166533],
       [-1.1097176 , -1.9102186 ]], dtype=float32)>


>We can assign "=" with assign (value), or assign_add (value) with "+ =", or assign_sub (value) with "-="

In [ ]:
new_value = tf.random.normal(shape=(2, 2))
a.assign(new_value)
for i in range(2):
    for j in range(2):
        assert a[i, j] == new_value[i, j]

In [ ]:
added_value = tf.random.normal(shape=(2,2))
a.assign_add(added_value)
for i in range(2):
    for j in range(2):
        assert a[i,j] == new_value[i,j]+added_value[i,j]

In [ ]:
tensor = tf.Variable([ [ [0., 1., 2.], [3., 4., 5.] ], [ [6., 7., 8.], [9., 10., 11.] ] ]) # tensor variable
print(tensor.shape)

(2, 2, 3)


In [ ]:
tensor1 = tf.reshape(tensor,[2,6]) # 2 rows 6 cols
tensor2 = tf.reshape(tensor,[1,12]) # 1 rows 12 cols
tensor1

<tf.Tensor: shape=(2, 6), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.,  4.,  5.],
       [ 6.,  7.,  8.,  9., 10., 11.]], dtype=float32)>

In [ ]:
tensor2 = tf.reshape(tensor,[1,12]) # 1 row 12 columns
tensor2

<tf.Tensor: shape=(1, 12), dtype=float32, numpy=
array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11.]],
      dtype=float32)>

In [ ]:
tf.rank(tensor)

<tf.Tensor: shape=(), dtype=int32, numpy=3>

>(the shape is () because the output here is a scalar value)

#### Specifying an element of a tensor

In [ ]:
tensor3 = tensor[1, 0, 2] # slice 1, row 0, column 2
tensor3

<tf.Tensor: shape=(), dtype=float32, numpy=8.0>

#### Casting a tensor to a NumPy/Python variable

In [ ]:
print(tensor.numpy())

[[[ 0.  1.  2.]
  [ 3.  4.  5.]]

 [[ 6.  7.  8.]
  [ 9. 10. 11.]]]


In [ ]:
print(tensor[1, 0, 2].numpy())

8.0


#### Finding the size (number of elements) of a tensor

In [ ]:
tensor_size = tf.size(input=tensor).numpy()
tensor_size

12

In [ ]:
#the datatype of a tensor
tensor3.dtype

tf.float32

### Tensorflow mathematical operations
>Can be used as numpy for artificial operations. Tensorflow can not execute these operations on the GPU or TPU.


In [ ]:
a = tf.random.normal(shape=(2,2))
b = tf.random.normal(shape=(2,2))
c = a+b
d = tf.square(c)
e = tf.exp(c)
print(a)
print(b)
print(c)
print(d)
print(e)

tf.Tensor(
[[-0.24671447  0.8228442 ]
 [ 1.4005157  -0.21337971]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-1.0893056  -0.98363787]
 [-0.5615219  -0.26913118]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-1.3360201  -0.16079366]
 [ 0.8389938  -0.4825109 ]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[1.7849498  0.0258546 ]
 [0.7039106  0.23281677]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[0.26288986 0.8514677 ]
 [2.3140376  0.61723167]], shape=(2, 2), dtype=float32)


### Performing element-wise primitive tensor operations

In [ ]:
tensor*tensor

<tf.Tensor: shape=(2, 2, 3), dtype=float32, numpy=
array([[[  0.,   1.,   4.],
        [  9.,  16.,  25.]],

       [[ 36.,  49.,  64.],
        [ 81., 100., 121.]]], dtype=float32)>

### Broadcasting
>Element-wise tensor operations support broadcasting in the same way that NumPy arrays do.

>The simplest example is that of multiplying a tensor by a scalar:

In [ ]:
tensor4 = tensor*4
print(tensor4)

tf.Tensor(
[[[ 0.  4.  8.]
  [12. 16. 20.]]

 [[24. 28. 32.]
  [36. 40. 44.]]], shape=(2, 2, 3), dtype=float32)


>the scalar multiplier 4 is—conceptually, at least—expanded into an array that can be multiplied element-wise with t2.

### Transpose Matrix multiplication

In [ ]:
matrix_u = tf.constant([[3,4,3]])
matrix_v = tf.constant([[1,2,1]])

tf.matmul(matrix_u, tf.transpose(a=matrix_v))

<tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[14]], dtype=int32)>

### Casting a tensor to another (tensor) datatype

In [ ]:
i = tf.cast(tensor1, dtype=tf.int32)
i

<tf.Tensor: shape=(2, 6), dtype=int32, numpy=
array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11]], dtype=int32)>

### With truncation

In [ ]:
j = tf.cast(tf.constant(4.9), dtype=tf.int32)
j

<tf.Tensor: shape=(), dtype=int32, numpy=4>

### Declaring Ragged tensors

`A ragged tensor is a tensor with one or more ragged dimensions. Ragged dimensions are dimensions that have slices that may have different lengths.There are a variety of methods for declaring ragged arrays, the simplest being a constant ragged
array.`

#### The following example shows how to declare a constant ragged array and the lengths of the individual slices:

In [ ]:
ragged =tf.ragged.constant([[5, 2, 6, 1], [], [4, 10, 7], [8], [6,7]])

print(ragged)
print(ragged[0,:])
print(ragged[1,:])
print(ragged[2,:])
print(ragged[3,:])
print(ragged[4,:])

<tf.RaggedTensor [[5, 2, 6, 1], [], [4, 10, 7], [8], [6, 7]]>
tf.Tensor([5 2 6 1], shape=(4,), dtype=int32)
tf.Tensor([], shape=(0,), dtype=int32)
tf.Tensor([ 4 10  7], shape=(3,), dtype=int32)
tf.Tensor([8], shape=(1,), dtype=int32)
tf.Tensor([6 7], shape=(2,), dtype=int32)


### Finding the squared difference between two tensors

In [ ]:
varx = [1,3,5,7,11]
vary = 5
varz = tf.math.squared_difference(varx,vary)
varz

<tf.Tensor: shape=(5,), dtype=int32, numpy=array([16,  4,  0,  4, 36], dtype=int32)>

>The Python variables, varx and vary, are cast into tensors and that vary is then broadcast across varx in this example. So, for example, the first calculation is (1-5)2 = 16.

#### Finding the mean

>The following is the signature of
>tf.reduce_mean().

`Note that this is equivalent to np.mean, except that it infers the return datatype from the input tensor,
whereas np.mean allows you to specify the output type (defaulting to float64):`

`tf.reduce_mean(input_tensor, axis=None, keepdims=None, name=None)`


In [ ]:
#Defining a constant
numbers = tf.constant([[4., 5.], [7., 3.]])

#### Find the mean across all axes (use the default axis = None)

In [ ]:
tf.reduce_mean(input_tensor=numbers)
#( 4. + 5. + 7. + 3.)/4 = 4.75

<tf.Tensor: shape=(), dtype=float32, numpy=4.75>

In [ ]:
tf.reduce_mean(input_tensor=numbers, axis=0) # [ (4. + 7. )/2 , (5. + 3.)/2 ] = [5.5, 4.]

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([5.5, 4. ], dtype=float32)>

In [ ]:
tf.reduce_mean(input_tensor=numbers, axis=0, keepdims=True)

<tf.Tensor: shape=(1, 2), dtype=float32, numpy=array([[5.5, 4. ]], dtype=float32)>

In [ ]:
tf.reduce_mean(input_tensor=numbers, axis=1) # [ (4. + 5. )/2 , (7. + 3. )/2] = [4.5, 5]

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([4.5, 5. ], dtype=float32)>

In [ ]:
tf.reduce_mean(input_tensor=numbers, axis=1, keepdims=True)

<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
array([[4.5],
       [5. ]], dtype=float32)>

In [ ]:
tf.random.normal(shape = (3,2), mean=10, stddev=2, dtype=tf.float32, seed=None, name=None)
ran = tf.random.normal(shape = (3,2), mean=10.0, stddev=2.0)
print(ran)

tf.Tensor(
[[11.012381   9.820841 ]
 [11.514592  11.424604 ]
 [ 8.9686985  7.321087 ]], shape=(3, 2), dtype=float32)


In [ ]:
tf.random.uniform(shape = (2,4), minval=0, maxval=None, dtype=tf.float32, seed=None, name=None)

<tf.Tensor: shape=(2, 4), dtype=float32, numpy=
array([[0.69793105, 0.12818623, 0.36479974, 0.6948261 ],
       [0.99294233, 0.607391  , 0.13492548, 0.45762825]], dtype=float32)>

#### Setting the seed

In [ ]:
tf.random.set_seed(11)
ran1 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
ran2 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
print(ran1) #Call 1
print(ran2)

tf.Tensor(
[[4 6]
 [5 2]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[9 7]
 [9 4]], shape=(2, 2), dtype=int32)


In [ ]:
tf.random.set_seed(11) #same seed
ran1 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
ran2 = tf.random.uniform(shape = (2,2), maxval=10, dtype = tf.int32)
print(ran1) #Call 2
print(ran2)

tf.Tensor(
[[4 6]
 [5 2]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[9 7]
 [9 4]], shape=(2, 2), dtype=int32)


In [ ]:
dice1 = tf.Variable(tf.random.uniform([10, 1], minval=1, maxval=7, dtype=tf.int32))
dice2 = tf.Variable(tf.random.uniform([10, 1], minval=1, maxval=7, dtype=tf.int32))
# We may add dice1 and dice2 since they share the same shape and size.
dice_sum = dice1 + dice2
# We've got three separate 10x1 matrices. To produce a single
# 10x3 matrix, we'll concatenate them along dimension 1.
resulting_matrix = tf.concat(values=[dice1, dice2, dice_sum], axis=1)
print(resulting_matrix)

tf.Tensor(
[[ 5  5 10]
 [ 4  3  7]
 [ 5  3  8]
 [ 3  3  6]
 [ 1  4  5]
 [ 4  1  5]
 [ 5  1  6]
 [ 6  4 10]
 [ 3  3  6]
 [ 2  3  5]], shape=(10, 3), dtype=int32)


In [ ]:
# 1-D tensor
t5 = tf.constant([2, 11, 5, 42, 7, 19, -6, -11, 29])
print(t5)


i = tf.argmax(input=t5)
print('index of max; ', i)
print('Max element: ',t5[i].numpy())


i = tf.argmin(input=t5,axis=0).numpy()
print('index of min: ', i)
print('Min element: ',t5[i].numpy())


t6 = tf.reshape(t5, [3,3])
print(t6)


i = tf.argmax(input=t6,axis=0).numpy() # max arg down rows
print('indices of max down rows; ', i)

i = tf.argmin(input=t6,axis=0).numpy() # min arg down rows
print('indices of min down rows ; ',i)
print(t6)

i = tf.argmax(input=t6,axis=1).numpy() # max arg across cols
print('indices of max across cols: ',i)

i = tf.argmin(input=t6,axis=1).numpy() # min arg across cols
print('indices of min across cols: ',i)


tf.Tensor([  2  11   5  42   7  19  -6 -11  29], shape=(9,), dtype=int32)
index of max;  tf.Tensor(3, shape=(), dtype=int64)
Max element:  42
index of min:  7
Min element:  -11
tf.Tensor(
[[  2  11   5]
 [ 42   7  19]
 [ -6 -11  29]], shape=(3, 3), dtype=int32)
indices of max down rows;  [1 0 2]
indices of min down rows ;  [2 2 0]
tf.Tensor(
[[  2  11   5]
 [ 42   7  19]
 [ -6 -11  29]], shape=(3, 3), dtype=int32)
indices of max across cols:  [1 0 2]
indices of min across cols:  [0 1 1]


In [ ]:
variable = tf.Variable([[1,3,5,7],[11,13,17,19]])
checkpoint= tf.train.Checkpoint(var=variable)
save_path = checkpoint.save('./vars')
variable.assign([[0,0,0,0],[0,0,0,0]])
variable
checkpoint.restore(save_path)
print(variable)

<tf.Variable 'Variable:0' shape=(2, 4) dtype=int32, numpy=
array([[ 1,  3,  5,  7],
       [11, 13, 17, 19]], dtype=int32)>


In [ ]:
a = tf.random.normal(shape=(2,2))
b = tf.random.normal(shape=(2,2))

with tf.GradientTape() as tape:
    tape.watch(a)
    c = tf.sqrt(tf.square(a)+tf.square(b))
    dc_da = tape.gradient(c,a)
    print(dc_da)

tf.Tensor(
[[-0.469929    0.89920384]
 [-0.66446555 -0.7976701 ]], shape=(2, 2), dtype=float32)


>For all variables, the calculation is tracked by default and used to find the gradient, so do not `usetape.watch()`

In [ ]:
a = tf.Variable(a)
with tf.GradientTape() as tape:
    c = tf.sqrt(tf.square(a)+tf.square(b))
    dc_da = tape.gradient(c,a)
    print(dc_da)

tf.Tensor(
[[-0.469929    0.89920384]
 [-0.66446555 -0.7976701 ]], shape=(2, 2), dtype=float32)


In [ ]:
with tf.GradientTape() as outer_tape:
    with tf.GradientTape() as tape:
        c = tf.sqrt(tf.square(a)+tf.square(b))
        dc_da = tape.gradient(c,a)
    d2c_d2a = outer_tape.gradient(dc_da,a)
    print(d2c_d2a)

tf.Tensor(
[[0.4264985  0.6867533 ]
 [0.44663432 0.50159544]], shape=(2, 2), dtype=float32)


In [ ]:
mnist = tf.keras.datasets.mnist
(train_x,train_y), (test_x, test_y) = mnist.load_data()

11490434/11490434 [==============================] - 1s 0us/step


In [ ]:
epochs=10
batch_size = 32 # 32 is default in fit method but specify anyway

>Next, normalize all the data points (x) to be in the float range zero to one, and of the float32 type.

>Also, cast the labels (y) to int64, as required:

In [ ]:
train_x, test_x = tf.cast(train_x/255.0, tf.float32), tf.cast(test_x/255.0, tf.float32)
train_y, test_y = tf.cast(train_y,tf.int64),tf.cast(test_y,tf.int64)

#### Building the Architecture

In [ ]:
mnistmodel1 = tf.keras.models.Sequential([
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(512,activation=tf.nn.relu),
tf.keras.layers.Dropout(0.2),
tf.keras.layers.Dense(10,activation=tf.nn.softmax)
])


#### Compiling the model

In [ ]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel1.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy', metrics = ['accuracy'])

#### Fitting the model

In [ ]:
mnistmodel1.fit(train_x, train_y, batch_size=32, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 10s 3ms/step - loss: 0.2206 - accuracy: 0.9344
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0978 - accuracy: 0.9704
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0709 - accuracy: 0.9771
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0540 - accuracy: 0.9826
Epoch 5/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.0439 - accuracy: 0.9860


#### Evaluate the mnistmodel1

In [ ]:
mnistmodel1.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0629 - accuracy: 0.9801


[0.06294650584459305, 0.9800999760627747]

>This represents a loss of 0.09 and an accuracy of 0.9801 on the test data.

>An accuracy of 0.98 means that out of 100 test data points, 98 were, on average, correctly identified by the model.

The second way to create a Sequential model
The alternative to passing a list of layers to the Sequential model's constructor is to use the add method, as follows, for the same architecture:



#### Building the Architecture & Compiling

In [ ]:
mnistmodel2 = tf.keras.models.Sequential();
mnistmodel2.add(tf.keras.layers.Flatten())
mnistmodel2.add(tf.keras.layers.Dense(512, activation='relu'))
mnistmodel2.add(tf.keras.layers.Dropout(0.2))
mnistmodel2.add(tf.keras.layers.Dense(10,activation=tf.nn.softmax))
mnistmodel2.compile (optimizer= tf.keras.optimizers.Adam(), loss='sparse_categorical_crossentropy',metrics = ['accuracy'])

#### Fitting the mnistmodel2

In [ ]:
mnistmodel2.fit(train_x, train_y, batch_size=64, epochs=5)

Epoch 1/5
938/938 [==============================] - 4s 3ms/step - loss: 0.2436 - accuracy: 0.9301
Epoch 2/5
938/938 [==============================] - 5s 5ms/step - loss: 0.1042 - accuracy: 0.9691
Epoch 3/5
938/938 [==============================] - 5s 6ms/step - loss: 0.0721 - accuracy: 0.9780
Epoch 4/5
938/938 [==============================] - 5s 6ms/step - loss: 0.0536 - accuracy: 0.9831
Epoch 5/5
938/938 [==============================] - 3s 3ms/step - loss: 0.0433 - accuracy: 0.9866


#### Evaluate the mnistmodel2

In [ ]:
mnistmodel2.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0640 - accuracy: 0.9790


[0.06400463730096817, 0.9789999723434448]

### The Keras functional API

`The functional API lets you build much more complex architectures than the simple linear stack of
Sequential models we have seen previously. It also supports more advanced models. These models
include multi-input and multi-output models, models with shared layers, and models with residual
connections.`

`Here is a short example, with an identical architecture to the previous two, of the use of the
functional API.`

In [ ]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(train_x,train_y), (test_x, test_y) = mnist.load_data()

train_x, test_x = train_x/255.0, test_x/255.0

epochs=10


#### Building the Architecture

In [ ]:
inputs = tf.keras.Input(shape=(28,28)) # Returns a 'placeholder' tensor
x = tf.keras.layers.Flatten()(inputs)
x = tf.keras.layers.Dense(512, activation='relu',name='d1')(x)
x = tf.keras.layers.Dropout(0.2)(x)
predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax, name='d2')(x)
mnistmodel3 = tf.keras.Model(inputs=inputs, outputs=predictions)

In [ ]:
optimiser = tf.keras.optimizers.Adam()
mnistmodel3.compile (optimizer= optimiser, loss='sparse_categorical_crossentropy', metrics = ['accuracy'])
mnistmodel3.fit(train_x, train_y, batch_size=32, epochs=epochs)

Epoch 1/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2217 - accuracy: 0.9344
Epoch 2/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0975 - accuracy: 0.9705
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0691 - accuracy: 0.9790
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0534 - accuracy: 0.9832
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0443 - accuracy: 0.9861
Epoch 6/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0370 - accuracy: 0.9879
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0302 - accuracy: 0.9897
Epoch 8/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0280 - accuracy: 0.9905
Epoch 9/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0252 - accuracy: 0.9914
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.022

#### Evaluate the mnistmodel3

In [ ]:
mnistmodel3.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0767 - accuracy: 0.9799


[0.0767001211643219, 0.9799000024795532]

In [ ]:
import tensorflow as tf

In [ ]:
class MNISTModel(tf.keras.Model):
    def __init__(self, num_classes=10):
        super(MNISTModel, self).__init__()
        # Define your layers here.
        inputs = tf.keras.Input(shape=(28,28)) # Returns a placeholder tensor
        self.x0 = tf.keras.layers.Flatten()
        self.x1 = tf.keras.layers.Dense(512, activation='relu',name='d1')
        self.x2 = tf.keras.layers.Dropout(0.2)
        self.predictions = tf.keras.layers.Dense(10,activation=tf.nn.softmax, name='d2')


    def call(self, inputs):
    # This is where to define your forward pass
    # using the layers previously defined in `__init__`
        x = self.x0(inputs)
        x = self.x1(x)
        x = self.x2(x)
        return self.predictions(x)

In [ ]:
mnistmodel4 = MNISTModel()

In [ ]:
batch_size = 32
steps_per_epoch = len(train_x)//batch_size
print(steps_per_epoch)
mnistmodel4.compile (optimizer= tf.keras.optimizers.Adam(), loss='sparse_categorical_crossentropy',metrics = ['accuracy'])
mnistmodel4.fit(train_x, train_y, batch_size=batch_size, epochs=epochs)

1875
Epoch 1/10
1875/1875 [==============================] - 9s 4ms/step - loss: 0.2212 - accuracy: 0.9348
Epoch 2/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.0962 - accuracy: 0.9702
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0698 - accuracy: 0.9778
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0521 - accuracy: 0.9830
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0418 - accuracy: 0.9865
Epoch 6/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0354 - accuracy: 0.9882
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0299 - accuracy: 0.9903
Epoch 8/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0286 - accuracy: 0.9907
Epoch 9/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0255 - accuracy: 0.9911
Epoch 10/10
1875/1875 [==============================] - 5s 3ms/step - loss: 

In [ ]:
mnistmodel4.evaluate(test_x, test_y)

313/313 [==============================] - 1s 2ms/step - loss: 0.0768 - accuracy: 0.9815


[0.07678413391113281, 0.9815000295639038]